### ISSUE TITLE
xUnit runs frequently into .NET EventListener bug

### ISSUE THREAD

**Issue by davkean** [xunit/xunit#646]
.NET has a bug in EventListener.DisposeOnShutdown: https://github.com/dotnet/corefx/issues/3793, that will be fixed in the next version of .NET Framework.

xUnit, unfortunately, runs into this bug ~ 2 or 3 times every 50 runs that we make, resulting in xUnit failing the run and outputting the following:

``` 
System.InvalidOperationException: Collection was modified; enumeration operation
 may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resour
ce)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at System.Diagnostics.Tracing.EventListener.DisposeOnShutdown(Object sender,
EventArgs e)
```

Understood that is probably only affecting us (Microsoft) at the moment and the bug exists on our side, but given that xUnit can release a lot quicker than .NET, would you consider a PR with a workaround that avoids the race?

**Comment by davkean** [comment]
tag @jaredpar.

**Comment by bradwilson** [comment]
I don't believe we're using event listener (or tracing at all) anywhere in our code. What is the proposed change?

**Comment by davkean** [comment]
We're not using it either, so we assumed xUnit was. I'll dig into it further and get back to you to with a proposed fix, if I find one.

**Comment by jaredpar** [comment]
I dug into it a bit and there appear to be four `EventSource` instances that get registered during the execution of an xunit.console.x86 process:

1. [FrameworkEventSource](http://referencesource.microsoft.com/#mscorlib/system/diagnostics/eventing/frameworkeventsource.cs,30a5f03a4a3ccc34): this is actually added twice.  Once from ResourceManager, and the other from using the thread pool.  The uses occur in different `AppDomain` instances and hence we get two events. 
2. [TplEtwProvider](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/TPLETWProvider.cs,e38cada0f78eafb7): triggered indirectly by Xunit.Sdk.TestFramework.Dispose()
3. [CDSCollectionEtwBCLProvider](http://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/CDSCollectionETWBCLProvider.cs,aea426773a7ea780,references): triggered by use of ConcurrentDictionary 

Hence I agree with @bradwilson thought that xunit isn't using EventSource directly here. 

I'm still a bit puzzled as to how xunit could be running into this CoreFx bug though.  The bug should in theory only reproduce if an event source is triggered during AppDomain tear down or during process exit.  If xunit triggered the EventSource asynchronously it would be easy to see how this could happen.  All the uses I see are synchronous.  

The only other piece of data we have is that this bug repros most often on unit test assemblies that run very fast (possibly by virtue of having 0 tests).  It doesn't seem to repro on any of our assemblies that take longer than 1 second to execute.  

In summary: I don't think there is really anything xunit can directly do here.  Just seems to accidentally running straight into the CoreFX bug in certain scenarios.

**Comment by bradwilson** [comment]
Resolving as External, please re-open if further investigation leads to needed changes in xUnit.net. Thanks.

**Comment by davkean** [comment]
Yep, thanks - don't think there's anything for xUnit to do here.
